state-chain/pallets/cf-swapping/src/migrations/swap_request_migration.rs

```diff
@@ -0,0 +1,106 @@
+use frame_support::traits::UncheckedOnRuntimeUpgrade;
+
+use crate::Config;
+
+use crate::*;
+use frame_support::pallet_prelude::Weight;
+#[cfg(feature = "try-runtime")]
+use sp_runtime::DispatchError;
+
+use codec::{Decode, Encode};
+
+pub mod old {
+	use super::*;
+	use cf_chains::{CcmDepositMetadata, ChannelRefundParameters, ForeignChainAddress};
+	use cf_primitives::{Asset, AssetAmount, Beneficiaries, SwapId};
+	use frame_support::Twox64Concat;
+
+	#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+	pub enum GasSwapState {
+		OutputReady { gas_budget: AssetAmount },
+		Scheduled { gas_swap_id: SwapId },
+		ToBeScheduled { gas_budget: AssetAmount, other_gas_asset: Asset },
+	}
+
+	#[allow(clippy::large_enum_variant)]
+	#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+	pub struct CcmState {
+		pub gas_swap_state: GasSwapState,
+		pub ccm_deposit_metadata: CcmDepositMetadata,
+	}
+
+	#[allow(clippy::large_enum_variant)]
+	#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+	pub enum SwapRequestState<T: Config> {
+		UserSwap {
+			ccm_deposit_metadata: Option<CcmState>,
+			output_address: ForeignChainAddress,
+			dca_state: DcaState,
+			broker_fees: Beneficiaries<T::AccountId>,
+		},
+		NetworkFee,
+		IngressEgressFee,
+	}
+
+	#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+	pub struct SwapRequest<T: Config> {
+		pub id: SwapRequestId,
+		pub input_asset: Asset,
+		pub output_asset: Asset,
+		pub refund_params: Option<ChannelRefundParameters>,
+		pub state: SwapRequestState<T>,
+	}
+
+	#[frame_support::storage_alias]
+	pub type SwapRequests<T: Config> =
+		StorageMap<Pallet<T>, Twox64Concat, SwapRequestId, SwapRequest<T>>;
+}
+
+pub struct SwapRequestMigration<T: Config>(PhantomData<T>);
+
+impl<T: Config> UncheckedOnRuntimeUpgrade for SwapRequestMigration<T> {
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+		Ok((old::SwapRequests::<T>::iter().count() as u64).encode())
+	}
+
+	fn on_runtime_upgrade() -> Weight {
+		crate::SwapRequests::<T>::translate_values::<old::SwapRequest<T>, _>(|old_swap_requests| {
+			Some(SwapRequest {
+				id: old_swap_requests.id,
+				input_asset: old_swap_requests.input_asset,
+				output_asset: old_swap_requests.output_asset,
+				refund_params: old_swap_requests.refund_params,
+				state: match old_swap_requests.state {
```

Comment:
It ocurred to me just now (missed this initially, sorry @albert-llimos) that the old state here in case of CCM contains `GasSwapState::ToBeScheduled { gas_budget, .. }` where `gas_budget` is what we set aside for swapping into the gas asset. As things stand, looks like this amount will be lost, and we probably want to put it back to `dca_state.remaining_input_amount` during the migration?

(I think `GasSwapState::ToBeScheduled` is going to be the only possible state/variant since we will have safe mode on, and if we get a vault swap during the upgrade it will be waiting in its initial state, which means the migration is much more straightforward.)

Reply:
I pushed this after your review so you hadn't even reviewed it ;) 
I thought about this. With the new logic the gas swap is done on the egress and CCM fees will be taking there (aka the new gas swap). I don't think it's much of a problem if we happen to swallow a CCM gas amount for a swap and then the egress fee that is taken is different. Also there is no CCM swaps going on and we will put safe mode, so as long it can't cause some failure I think it's fine as it is.

Reply:
If I understand correctly, it is not a matter of charging a different fee, but essentially charging the fee twice. Sure, we would have to be unlucky to hit this, but it also seems very simple to just add the gas amount back to input. Not a strong opinion, I will leave it to you to decide whether it is worth it or not.

Reply:
We also then need to change the state of another storage slot (`SwapState`), I think it's just unnecessary. This would only happen in a CCM+DCA before the first chunk of DCA, since the gas swap is scheduled after that. Besides the main point that we have never even had a swap like that. The downside is just charging egress fees twice in that extremely rare scenario, which is fine.

Reply:
I agree, I think we can live with this.